### ISSUE TITLE
Enable history purging when not using mysql

### ISSUE THREAD

**Issue by jl98800** [HubSpot/Singularity#1694]
Version of Singularity:

Singularity 0.16.2 and 0.14.1
Expected behavior

The expected behavior is for singularity to not time out and restart - which restarts all running tasks.
Actual behavior

java.lang.IllegalStateException: Timed out waiting response - singularity restarts, and restarts all running tasks (bad for business).
Steps to reproduce behavior

Introduce a heavy load (lots active requests/tasks - like 10,000 active requests and over 50 active tasks.

Other helpful info and things we've tried to do. Please provide specifics to using a DB to manage history (for purging).

Found this in the logs - this happens consistently every thirty minutes:

java.lang.IllegalStateException: Timed out waiting response for objects from /tasks/history, waited 5000 millis

This can be caused by a large amount of request or task history, a heavy task load in singularity, or problems with zookeeper.

Updating singularity.yaml in a custom docker image can help (Audience Modeling is doing this). There are several parameters to consider here: https://github.com/HubSpot/Singularity/blob/master/Docs/reference/configuration.md#history-purging

Or this configuration in singularity.yaml:

historyPurging:
enabled: true
deleteTaskHistoryAfterDays: 7

Other parameters to consider besides deleteTaskHistoryAfterDays:

deleteTaskHistoryAfterTasksPerRequest  (by record count)
checkTaskHistoryEveryHours

The other option is to spread out the load/requests in Singularity.

Lastly, cleaning up the singularity state/app/scheduled tasks is another option (manually through the zookeeper UI (aka exhibitor). Though in my experience this has limited success with this sort of issue.

stderr shows ! java.lang.IllegalStateException: Timed out waiting response for objects from /racks, waited 5000 millis

it was happening because they were submitting lots of simultaneous tasks and each of them had lots of task history which singularity fetches

AMD team should try to throttle task scheduling from their end because there is no 'fix' for singularity other than deleting all task history. There are parameters we can put into singularity.yaml to clear history at a certain threshold.

AMD team are running tasks allocated 280+GB of memory

i think part of the issue here is that they have 8055 'requests' marked 'active' every time a new 'task' runs, singularity queries all of that history.

Tried to manage this with the following entry in singularity.yaml:

historyPurging:
enabled: true
deleteTaskHistoryAfterDays: 7
and
deleteTaskHistoryAfterTasksPerRequest: 2000

Do these parameters require the use of a DB to be effective???

**Comment by ssalinas** [comment]
A few things in response to this:

- Those timeouts are timeouts querying zookeeper. If you want to give it more time in those you can set `zookeeper.connectTimeoutMillis` to a higher value in your yaml
- Singularity should _never_ be restarting all running tasks when restarting. I'm curious of what you are doing here. We restart Singularity deploying new versions multiple times a day and have not seen it ever restart all tasks. Can you provide more information on what you see in this case?
- In later versions of Singularity we have introduced a number performance improvements that require less queries to zookeeper. More information is cached in memory so that it can handle incoming offers, new tasks, and status updates more quickly without putting excessive load on zookeeper
- History purging will only be effective with a DB currently. With mysql set up, data for inactive tasks is offloaded to mysql from zookeeper.
- Cleaning up scheduled tasks manually in zookeeper has a high likelihood of corrupting Singularity's internal state (part of why we abort and expect to restart when we cannot contact zookeeper).

So in short I would suggest lengthening the zookeeper timeout via the parameter mentioned above as well as upgrading to a newer version of Singularity. I would also love more details around the restarting piece as that is something that should never happen

**Comment by jl98800** [comment]
Stephen,

Thanks for your detailed reply.  I may be mistaken about all running tasks
being restarted.  Getting some info 2nd hand.

Here's the current setting for connectTimeoutMillis - which isn't high
enough according the restart in the logs "waited 5000".  What should the
number be for a VERY busy system?

zookeeper:
  quorum: localhost:2181
  zkNamespace: singularity
  sessionTimeoutMillis: 600000
  connectTimeoutMillis: 60000
  retryBaseSleepTimeMilliseconds: 1000
  retryMaxTries: 3

Also, can you provide documentation to set up MySQL DB to use with
Singularity?

Thanks,
John

**Comment by ssalinas** [comment]
We have a reference [in the docs](http://getsingularity.com/Docs/reference/database.html) for setting up Singularity with mysql. As for the `connectTimeoutMillis`, I'd say starting with something like 10000 or 15000 would be good.

I'll leave this issue open as a TODO for us to also add history purging when only using zookeeper. When getting to thousands of tasks it could definitely slow things down

**Comment by jl98800** [comment]
Hi

We have our connection timeout set to 60,000. It only seems to delay the inevitable. 

Thanks for working on history purging!

Thanks,

John

**Comment by ssalinas** [comment]
Relized we do have a configuration setting for this in `deleteTasksFromZkWhenNoDatabaseAfterHours`, which should run the history purger on zk. Closing this one
